[Merged by Bors] - feat(data/fintype): pigeonhole principles

kmill opened the pull request:
Add pigeonhole principles for finitely and infinitely many pigeons in finitely many holes. There are also strong versions of these, which say that there is a hole containing at least as many pigeons as the average number of pigeons per hole. Fixes #2272.

---
<!-- put comments you want to keep out of the PR commit here -->

I'm not sure what they should be called or where they should go.  Also, the proof of `fintype.strong_pigeonhole` can surely be simplified!

kmill committed:
Added pigeonhole principles for finitely and infinitely many pigeons

Also added strong pigeonhole principles: there is a preimage whose
cardinality is at least as large as the average cardinality.

kmill committed:
Cleaning up the strong pigeonhole principle proof

kmill committed:
Cleaned up strong pigeonhole proof even more

Also added finset.univ_nonempty to give a proof of univ.nonempty

kmill committed:
Added finset versions of pigeonhole principles

eric-wieser commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,57 @@ open opposite
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
+-/
+lemma strong_pigeonhole {s : finset α} {t : finset β} (hne : t.nonempty) [decidable_eq β]
```
The lemma is true even without `hne`, isn't it?

kmill replied:
If both `s` and `t` are empty, then the remaining hypotheses can be trivially satisfied, but the conclusion would imply `t` is nonempty, a contradiction.  Putting a constraint on `t` being nonempty is enough to prevent this situation from happening.

eric-wieser replied:
Hmm, I suppose you end up needing to prove `∃ y ∈ t, true`, which requires `t` to be non-empty.

kmill committed:
Added some finset lemmas to simplify proof more

These lemmas might be too specialized, but they help relate filter and
image a little more.

eric-wieser commented on src/data/finset/basic.lean:
```diff
@@ -1362,6 +1362,18 @@ by simp only [mem_def, image_val, mem_erase_dup, multiset.mem_map, exists_prop]
 theorem mem_image_of_mem (f : α → β) {a} {s : finset α} (h : a ∈ s) : f a ∈ s.image f :=
 mem_image.2 ⟨_, h, rfl⟩
 
+lemma filter_mem_image_eq_image (f : α → β) (s : finset α) (t : finset β) (h : ∀ x ∈ s, f x ∈ t) :
```
Does adding `@[simp]` to this lemma make sense?

eric-wieser replied:
Also, would it be more or less useful to state `h` as `t ⊇ s.image f`?

kmill replied:
I originally had it as a simp lemma, but I think it's never applicable because of that `h` hypothesis.

It's possible to state it using `t ⊇ s.image f`, but it seems like this is a fact that's not too easy to use directly except through `image_subset_iff`.  (The precedent I was following was the `hf` argument to `card_le_card_of_inj_on`.)

eric-wieser commented on src/data/fintype/basic.lean:
```diff
@@ -434,6 +437,17 @@ lemma fintype.card_le_of_injective [fintype α] [fintype β] (f : α → β)
+  classical, by_contra hf, push_neg at hf,
+  refine nat.lt_le_antisymm h (fintype.card_le_of_injective f _),
+  intros x y, contrapose, apply hf,
+end
```
Can this be written in terms of `finset.pigeonhole`, or vice versa?

Might be nice to have the doc-comments reference the other lemma.

kmill replied:
Yeah, it's possible to write `fintype.pigeonhole` in terms of `finset.pigeonhole`:
```lean
lemma fintype.pigeonhole' [fintype α] [fintype β] (f : α → β) (h : fintype.card β < fintype.card α) :
  ∃ x y, x ≠ y ∧ f x = f y :=
by { obtain ⟨x, _, y, _, h⟩ := @finset.pigeonhole α β univ univ h f (by simp),
     exact ⟨x, y, h⟩ }
```
I just figured it was a trivial application of `fintype.card_le_of_injective`, so it doesn't hurt to have independent proofs, but I might change it.

I agree it would be nice to have doc comments -- I'll be sure to add them once it's clear what these lemmas should be called and where they should go.

kmill replied:
Changed it and added some documentation

kim-em commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,50 @@ open opposite
+  have h' := lt_of_lt_of_le h hn,
+  exact nat.lt_asymm h' h',
+
+  have key : ∀ y ∈ t, (s.filter (λ x, f x = y)).card ≠ 0 → y ∈ s.image f,
```
This seems a generally useful fact, so should be proved separately.

kmill replied:
:+1: added `finset.preimage_card_ne_zero_iff_mem_image`

kim-em commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,50 @@ open opposite
+lemma strong_pigeonhole {s : finset α} {t : finset β} (hne : t.nonempty) [decidable_eq β]
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t)
+  (n : ℕ) (hn : t.card * n ≤ s.card) :
+  ∃ y ∈ t, n ≤ (s.filter (λ x, f x = y)).card :=
```
I feel like this is not actually the strongest statement, and you should instead assume `t.card * n < s.card` and conclude `n < (s.filter (λ x, f x = y)).card`.

With this variant, you no longer need a `nonempty` assumption, either.

kmill replied:
Oh right... I intended for `n` to be the ceiling of `s.card / t.card` and kept assuming that's what I did!

(Convincing myself you have the strongest statement: for `n` to be the ceiling, it would be `hn : t.card * n ≤ s.card + t.card - 1`, and with `n = n' + 1`, this becomes `hn : t.card * n' ≤ s.card - 1`, which can safely be written as `hn : t.card * n' < s.card`.)

kmill replied:
Updated it to the stronger statement.  This makes it so `finset.univ_nonempty` is not used by this PR -- should I leave it in since it still seems useful to have around?

kim-em commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,50 @@ open opposite
+namespace finset
+
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
```
Once we agree on the right statement, let's state it completely in the doc-string.

kmill replied:
Stated.

kmill committed:
Made strong pigeonhole principle stronger, added more documentation

bryangingechen commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,62 @@ open opposite
+
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
+This is the `fintype` version of `finset.strong_pigeonhole`.
```
Can you copy the longer description from `finset.strong_pigeonhole` here, to keep this self-contained?

kmill replied:
Done

kmill committed:
Copied description to `fintype.pigeonhole`

kmill committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into pigeonhole

kmill committed:
Removed unused arguments, simplified a proof thanks to Yury

urkud commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
+setting `n` so that `n + 1` is at most the ceiling of `s.card / t.card`,
+then there is an element of `t` whose preimage contains more
+than `n` elements.  (We formulate the constraint on `n` as
+`t.card * n < s.card`, which incorporates that `s` and `t` must be nonempty.)
```
This inequality implies only `s.nonempty`. You need `hf` to deduce `t.nonempty`.

kmill replied:
I meant that in the context of there being a function between finite sets, but I can clarify this if you want.

urkud commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
+setting `n` so that `n + 1` is at most the ceiling of `s.card / t.card`,
+then there is an element of `t` whose preimage contains more
+than `n` elements.  (We formulate the constraint on `n` as
+`t.card * n < s.card`, which incorporates that `s` and `t` must be nonempty.)
```
This version deals with `fintype`s, not `finset`s.

robertylewis commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
 end with_top
+
+namespace finset
+
```
Is there a reason these are in the `algebra/big_operators/order` file? Their presence should be clarified in the module doc if there's a reason -- I would never expect to find them here. But maybe they belong somewhere else.

kmill replied:
They're only here for my own convenience: the proof uses big operators and `sum_le_sum`.  It's seemed to me they ought to be elsewhere, but I'm not sure where that might be!  Maybe create `combinatorics/basic`?

robertylewis commented on src/data/finset/basic.lean:
```diff
@@ -1682,6 +1696,19 @@ begin
+If there are more pigeons than pigeonholes, then there are two pigeons
+in the same pigeonhole.
+-/
+lemma pigeonhole {s : finset α} {t : finset β} (hc : t.card < s.card)
```
`mathlib` generally avoids declaration names like `pigeonhole`. It's better to follow the standard naming convention and make sure that "Pigeonhole principle" appears at the beginning of the doc string. As for what the mathlib name of this should be, I'm not sure, maybe better to get some opinions here.

kmill replied:
Yeah, I wasn't sure what the mathlib name should be, so I left it with the common name as a placeholder.  I think for a little bit I had `exists_distinct_of_card_lt`, but that didn't seem right.

robertylewis commented on src/data/finset/basic.lean:
```diff
@@ -1682,6 +1696,19 @@ begin
+in the same pigeonhole.
+-/
+lemma pigeonhole {s : finset α} {t : finset β} (hc : t.card < s.card)
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t) :
```
```suggestion
  {f : α → β} (hf : ∀ a ∈ s, f a ∈ t) :
```

robertylewis commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
+-/
+lemma strong_pigeonhole {s : finset α} {t : finset β} [decidable_eq β]
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t)
+  (n : ℕ) (hn : t.card * n < s.card) :
```
```suggestion
  {f : α → β} (hf : ∀ a ∈ s, f a ∈ t)
  {n : ℕ} (hn : t.card * n < s.card) :
```

robertylewis commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
+  classical, by_contra hz, push_neg at hz,
+  suffices h : s.card ≤ t.card * n,
+  have h' := lt_of_le_of_lt h hn,
+  exact nat.lt_asymm h' h',
```
```suggestion
  { have h' := lt_of_le_of_lt h hn,
    exact nat.lt_asymm h' h' },
```
This lets us see where the new goal begins and ends.

kmill committed:
Incorporating review suggestions (still need better names and locations!)

kmill committed:
Merge remote-tracking branch 'origin/master' into pigeonhole

kmill committed:
Fixed liner error

kmill committed:
forgot to remove some duplicated arguments

urkud commented on src/algebra/big_operators/order.lean:
```diff
@@ -308,3 +309,70 @@ open opposite
+t.card`, then there is an element of `t` whose preimage contains more
+than `n` elements.  (We formulate the constraint on `n` as
+`t.card * n < s.card`.  Since we have a function `f` from `s` to `t`,
+this implies `t.card ≠ 0`, so `s.card / t.card` is defined.)
```
This lemma is about `fintype`s while the docstring is about finite sets.

kmill committed:
fixed copy/paste of docstring to refer to fintype rather than finset

kmill committed:
moved strong pigeonhole principles to combinatorics/pigeonhole

kmill committed:
removed extraneous import

urkud committed:
Merge branch 'master' into pigeonhole

urkud committed:
Golf, add versions with weights

urkud committed:
Drop duplicate args, add a copyright header

kmill committed:
Copyright notice, documentation, golfing, linting

kmill committed:
merge

urkud committed:
Merge branch 'master' into pigeonhole

urkud committed:
Rename, add more versions (now we have 20 thms in the file)

kmill committed:
fixed two names in data.finset.basic

urkud commented on src/data/fintype/basic.lean:
```diff
@@ -1121,6 +1133,45 @@ assume (hf : injective f),
+
+See also: `fintype.infinite_pigeonhole`
+-/
+lemma fintype.strong_infinite_pigeonhole [infinite α] [fintype β] (f : α → β) :
```
```suggestion
See also: `fintype.exists_ne_map_eq_of_infinite`
-/
lemma fintype.exists_infinite_fiber [infinite α] [fintype β] (f : α → β) :
```

urkud commented on src/data/fintype/basic.lean:
```diff
@@ -1121,6 +1133,45 @@ assume (hf : injective f),
+See also: `fintype.pigeonhole`, `fintype.strong_infinite_pigeonhole`
+-/
+lemma fintype.infinite_pigeonhole [infinite α] [fintype β] (f : α → β) :
+  ∃ x y : α, x ≠ y ∧ f x = f y :=
```
```suggestion
See also: `fintype.exists_ne_map_eq_of_card_lt`, `fintype.exists_infinite_fiber`.
-/
lemma fintype.exists_ne_map_eq_of_infinite [infinite α] [fintype β] (f : α → β) :
  ∃ x y : α, x ≠ y ∧ f x = f y :=
```

urkud reviewed (commented):
Forgot to rename 2 lemmas.

urkud committed:
Apply suggestions from code review

kim-em reviewed (approved):
Looks good to me! Let's merge.
